### PR TITLE
ASR: Support Array storage order

### PR DESCRIPTION
**Pull request by Shaikh-Ubaid**
towards #1249

### CODE REVIEW

File: src/libasr/ASR.asdl [hunk 1]
```diff
@@ -273,8 +273,8 @@ expr
 
     | Var(symbol v)
 
-    | ArrayConstant(expr* args, ttype type)
-    | ArrayItem(expr v, array_index* args, ttype type, expr? value)
+    | ArrayConstant(expr* args, ttype type, arraystorage storage_format)
+    | ArrayItem(expr v, array_index* args, ttype type, arraystorage storage_format, expr? value)
     | ArraySection(expr v, array_index* args, ttype type, expr? value)
     | ArraySize(expr v, expr? dim, ttype type, expr? value)
     | ArrayBound(expr v, expr? dim, ttype type, arraybound bound,
```
Comment (Shaikh-Ubaid) [inline src/libasr/ASR.asdl:280]: Currently, the `arraystorage` was being needed in the `ArrayItem` in the `wasm` backend. Hence, I added the attribute `arraystorage` only to `ArrayItem` and also `ArrayConstant`.

Comment (Shaikh-Ubaid) [reply src/libasr/ASR.asdl:280]: It seems it is possible that the `arraystorage` might be needed in others like `ArraySection`. I guess, we can add it as the need arises.

**Comment by Shaikh-Ubaid** [comment]
This is ready. Please possibly review and please share feedback.

**Comment by Shaikh-Ubaid** [comment]
I will send a `PR` to `LFortran` with these changes once this is complete/merged.

**Comment by certik** [comment]
It seems it should rather be part of the type. Here is one way to do it cleanly:
```
--- a/src/libasr/ASR.asdl
+++ b/src/libasr/ASR.asdl
@@ -327,22 +327,23 @@ expr
 --     integer, so we also use kind=4 for the default logical.)
 
 ttype
-    = Integer(int kind, dimension* dims)
-    | Real(int kind, dimension* dims)
-    | Complex(int kind, dimension* dims)
-    | Character(int kind, int len, expr? len_expr, dimension* dims)
-    | Logical(int kind, dimension* dims)
+    = Integer(int kind)
+    | Real(int kind)
+    | Complex(int kind)
+    | Character(int kind, int len, expr? len_expr)
+    | Logical(int kind)
     | Set(ttype type)
     | List(ttype type)
     | Tuple(ttype* type)
-    | Struct(symbol derived_type, dimension* dims)
-    | Enum(symbol enum_type, dimension *dims)
-    | Union(symbol union_type, dimension *dims)
-    | Class(symbol class_type, dimension* dims)
+    | Struct(symbol derived_type)
+    | Enum(symbol enum_type)
+    | Union(symbol union_type)
+    | Class(symbol class_type)
     | Dict(ttype key_type, ttype value_type)
     | Pointer(ttype type)
+    | Array(ttype type, dimension* dims, arraystorage storage_format)
     | CPtr()
-    | TypeParameter(identifier param, dimension* dims)
+    | TypeParameter(identifier param)
 
 restriction_arg = RestrictionArg(identifier restriction_name, symbol restriction_func)
 
```
But that requires more refactoring, so we can do it in a separate PR.

**Review by certik — APPROVED** [review]
I think this is good for now, but we'll have to later refactor how we handle arrays and move this into the type itself I think.

**Pull request merged**